hcl/eval/partial.go

```diff
@@ -0,0 +1,1386 @@
+		e.emit()
+	}
+}
+
+func (e *engine) emitComments() {
+	for e.hasTokens() && e.peek().Type == hclsyntax.TokenComment {
+		e.emit()
+	}
+}
+
+func (e *engine) emitnlparens() {
+	if e.multiline > 0 {
+		e.emitnl()
+	}
+}
+
+func (e *engine) skipTokens(from int, tokens ...hclsyntax.TokenType) int {
+	i := from
+	for e.hasTokens() {
+		found := false
+		for _, t := range tokens {
+			if e.peekn(i).Type == t {
+				found = true
+				break
+			}
+		}
+
+		if !found {
+			break
+		}
+
+		i++
+	}
+	return i
+}
+
+func (e *engine) skip(from int) int {
+	if e.multiline > 0 {
+		return e.multilineSkip(from)
+	} else {
+		return e.skipTokens(from, hclsyntax.TokenComment)
+	}
+}
+
+func (e *engine) multilineSkip(from int) int {
+	return e.skipTokens(from, hclsyntax.TokenNewline, hclsyntax.TokenComment)
+}
+
+func (e *engine) evalExpr() error {
+	_, thisNode := e.newnode()
+
+loop:
+	for {
+		e.emitnlparens()
+		e.emitComments()
+		switch t := e.peek().Type; {
+		case isUnaryOp(t):
+			thisNode.hasOp = true
+			e.emit()
+		default:
+			break loop
+		}
+	}
+
+	e.emitnlparens()
+	beginPos := e.pos
+	tok := e.peek()
+	// exprTerm
+	switch tok.Type {
+	case hclsyntax.TokenEOF:
+		e.emit()
+	case hclsyntax.TokenOHeredoc:
+		e.emit()
+
+		for e.hasTokens() &&
+			e.peek().Type != hclsyntax.TokenCHeredoc &&
+			e.peek().Type != hclsyntax.TokenEOF { // TODO(i4k): hack to imitate hashicorp lib
+			e.emit()
+		}
+		if !e.hasTokens() {
+			panic("expect close heredoc")
+		}
+
+		e.emit()
+	case hclsyntax.TokenOQuote:
+		err := e.evalString()
+		if err != nil {
+			return err
+		}
+		e.commit()
+	case hclsyntax.TokenIdent:
+		switch string(tok.Bytes) {
+		case "true", "false", "null":
+			e.emit()
+		default:
+			if e.canEvaluateIdent() {
+				err := e.evalIdent()
+				if err != nil {
+					return err
+				}
+				e.commit()
+
+			} else {
+				e.emit()
+			}
+		}
+	case hclsyntax.TokenOParen:
+		e.emit()
+		e.emitnl()
+
+		e.multiline++
+
+		err := e.evalExpr()
+		if err != nil {
+			return err
+		}
+
+		e.commit()
+		e.emitnl()
+
+		if e.peek().Type != hclsyntax.TokenCParen {
+			panic(e.peek().Type)
+		}
+
+		e.emit()
+		e.multiline--
+	case hclsyntax.TokenOBrace, hclsyntax.TokenOBrack:
+		var err error
+
+		var closeToken hclsyntax.TokenType
+
+		openToken := tok.Type
+		if openToken == hclsyntax.TokenOBrace {
+			closeToken = hclsyntax.TokenCBrace
+		} else {
+			closeToken = hclsyntax.TokenCBrack
+		}
+
+		pos := e.multilineSkip(1)
+		next := e.peekn(pos)
+		switch {
+		case isForExpr(next):
+			err = e.evalForExpr(openToken, closeToken)
+		case openToken == hclsyntax.TokenOBrace:
+			err = e.evalObject()
+		case openToken == hclsyntax.TokenOBrack:
+			err = e.evalList()
+		default:
+			panic("unexpected")
+		}
+
+		if err != nil {
+			return err
+		}
+
+		e.commit()
+
+	case hclsyntax.TokenNumberLit:
+		e.emit()
+	}
+
+	if e.pos == beginPos {
+		panic(sprintf("bug: no advance in the position: %s (%s)", e.peek().Type, e.tokens[e.pos:].Bytes()))
```

Comment:
```suggestion
		panic(sprintf("can't advance in the position: %s (%s)", e.peek().Type, e.tokens[e.pos:].Bytes()))
```

Or we could add "bug:" on all panics, but I think it is assumed that when a panic happens it is a bug :laughing: